### DISCUSSION TITLE
How to Import .so File in MAUI

### DISCUSSION THREAD

**Discussion by Yxxxxl** [dotnet/maui#12349] [Q&A] [answered]
Halo,can anyone help,how to import a so file in MAUI app?

I use DllImport to import .so file. But it throw System.DllNotFoundException: 'libadd.so'.

`[DllImport("libadd.so", EntryPoint = "add")]`

My  .so file store path is Resources/lib/x86_64/libadd.so

Thank you so much~

**Comment by xtuzy** [comment, marked as answer]
Use Android Class library store .so, 

example;https://github.com/xtuzy/Veldrid.SPIRV.NativeAsset/tree/master/Veldrid.SPIRV.AndroidNativeAsset

doc of xamarin:https://learn.microsoft.com/en-us/xamarin/cross-platform/cpp/

**Reply by Yxxxxl** [reply]
thanks，now i have a new problem.My apps can run in x86_64 well,but it cannot find .so when i run in arm64.

**Reply by 406345** [reply]
so as my . Library can not be found on my ARM64 device...

**Reply by Yxxxxl** [reply]
You can check your .so File 's  File Properties. If you already to set build action to AndroidNativeLibrary.

**Reply by 406345** [reply]
> You can check your .so File 's File Properties. If you already to set build action to AndroidNativeLibrary.

i did not have that option，i modify the project file directly . there are my so files in the apk in lib folder but dllimport says so not found. it is very weird.......

**Reply by xtuzy** [reply]
If you use Android Class Library(TargetFramework is net6.0-android),AndroidNativeLibrary is default:
![image](https://user-images.githubusercontent.com/17793881/211131089-19dfcfaf-91ab-4e86-b5e8-6221ceb8431a.png)

If you apk have .so, it's path should like this:
![image](https://user-images.githubusercontent.com/17793881/211130995-74a6031c-c1fa-47a8-8ea5-cb794f94cfa3.png)
DllImport like this:
![image](https://user-images.githubusercontent.com/17793881/211131195-a1237435-f4b9-400a-b2fd-8134cfea58b3.png)

**Reply by Xyncgas** [reply]
Are you ok in china, I haven't seen you around for a while

**Reply by xtuzy** [reply]
i'm ok🤗, thank you for your concern. i spend some time at https://github.com/xtuzy/Veldrid.Samples, don't see discussion of maui often.
